### ISSUE TITLE
Splitting the Geometry

### ISSUE THREAD

**Issue by rupakraj** [DotSpatial/DotSpatial#1029]
Hello,
I am trying to make an application to split the Polygon (Geometry) which is stored on the PostgreSql as 
 `polygon geometry(Geometry,3148),` using PostGIS extension. I loaded the same polygon into the DotSpatial map by following technique
`
 FeatureSet fs = new FeatureSet(FeatureType.Polygon);
IGeometry geom = GEOMETRY_FROM_DB;
IFeature feature = fs.AddFeature(geom);
var layer = map1.Layers.Add(fs);
`
Now my goal is to split the `geom` object using different polygon splitting techniques such as "Offset distance method with reference to selected baseline".

I have computed the offset line parallel to the base line using `NetTopologySuite.Operation.Buffer.OffsetCurveBuilde`. 

![image](https://user-images.githubusercontent.com/1549123/28496374-d673ece2-6f88-11e7-8b2b-0d1f4ed8cb16.png)

My Line 2 is `LineSegment` Now I wants to extend this line to cut the polygon and return the resulting polygons.

Can anyone help me to solve this problem?

**Comment by jany-tenaj** [comment]
Try [Geometry.Intersection](https://github.com/DotSpatial/NetTopologySuite/blob/ds_master/NetTopologySuite/Geometries/Geometry.cs#L1596) like `geom.Intersection(line2)`.

The parts of the result should be the two polygons you want.

**Comment by rupakraj** [comment]
Thanks @jany-tenaj for replay. I got a result but of type `LineString`. My problem is `line2` is not completely passed by cutting the polygon. In the figure, `Red` colored part is `line2` and extended(black thin line) is just extension of line 2. which actually does not exits.

**Comment by jany-tenaj** [comment]
I think you could use [Scalar Multiplication](https://en.wikipedia.org/wiki/Scalar_multiplication) to make your line2 long enough so that it cuts the polygon on the other end too.

**Comment by rupakraj** [comment]
How can I extend the line into infinity? I tried with vector class as well but seems bit difficult to work.

**Comment by jany-tenaj** [comment]
No idea. What for would one need an infinit line?

**Comment by rupakraj** [comment]
I was thinking about infinite line just to ensure that it cuts the polygon. While exploring I have seen `ClipPolygonWithLine` class. Can I clip that polygon using line which is not completely passed through polygon?

`bool IsClipped = ClipPolygonWithLine.AccurateClipPolygonWithLine(ref polygon, ref line,ref  splittedPolygon;`

**Comment by jany-tenaj** [comment]
No idea what result you get with lines which do not pass through completely. But I don't think the function will make them longer so you might get the same result as above. 

I think you can make sure that your line is long enough the following way.

- Get the Envelope of the line. 
- Get the angle of the line. (I'm not sure whether Linestring.Angle returns the angle that shows the direction of the line, so you might have to use [RadAngle ](https://github.com/DotSpatial/DotSpatial/blob/4c9e850c28ecbe0e8fbe48e00027ef2a0cec8d04/Source/DotSpatial.NTSExtension/LineStringExt.cs#L48)and convert it to degree)
- Check whether the line goes bottom up, left right, right left, or upside down.
- According to the way the line goes get the yMax, xMay, xMin or yMin value of the Envelope.
- Add (max values) / subtract (min values) 1 to make sure you're truely larger than the Envelope.
- Devide the resulting value by the vector value (xValues through x, yValues through y)
- Round the resulting value up (positive values) or down (negative values)
- Use the resulting value as scalar multplicator

**Comment by rupakraj** [comment]
Thanks @jany-tenaj for valuable comments. I solved the problem by extending the line using its coordinates and angle. I extended the endpoints of the line using same angle(slope) upto the perimeter of the polygon to ensure it cuts the polygon. I used simple `sin` `cos` rules and added the offset to calculated values.

```
xExtended = x+length*cos(angle)
yExtended = y+length*sin(angle)
```
Output I got is something like below

![image](https://user-images.githubusercontent.com/1549123/28522616-737f7bfe-7098-11e7-97df-fb917ac833e1.png)
